controllers/tlspolicy_status_updater.go

```diff
@@ -0,0 +1,233 @@
+
+		conditions = issuer.Status.Conditions
+	case certmanagerv1.ClusterIssuerKind:
+		objs := topology.Objects().Children(gw)
+		obj, ok := lo.Find(objs, func(o machinery.Object) bool {
+			return o.GroupVersionKind().GroupKind() == CertManagerClusterIssuerKind && o.GetName() == tlsPolicy.Spec.IssuerRef.Name
+		})
+		if !ok {
+			err := fmt.Errorf("%s \"%s\" not found", tlsPolicy.Spec.IssuerRef.Kind, tlsPolicy.Spec.IssuerRef.Name)
+			logger.Error(err, "error finding object in topology")
+			return err
+		}
+
+		issuer := obj.(*controller.RuntimeObject).Object.(*certmanagerv1.ClusterIssuer)
+		conditions = issuer.Status.Conditions
+	default:
+		return fmt.Errorf(`invalid value %q for issuerRef.kind. Must be empty, %q or %q`, tlsPolicy.Spec.IssuerRef.Kind, certmanagerv1.IssuerKind, certmanagerv1.ClusterIssuerKind)
+	}
+
+	transformedCond := utils.Map(conditions, func(c certmanagerv1.IssuerCondition) metav1.Condition {
+		return metav1.Condition{Reason: c.Reason, Status: metav1.ConditionStatus(c.Status), Type: string(c.Type), Message: c.Message}
+	})
+
+	if !meta.IsStatusConditionTrue(transformedCond, string(certmanagerv1.IssuerConditionReady)) {
+		return fmt.Errorf("%s not ready", tlsPolicy.Spec.IssuerRef.Kind)
+	}
+
+	return nil
+}
+
+func (t *TLSPolicyStatusUpdaterReconciler) isCertificatesReady(ctx context.Context, p machinery.Policy, topology *machinery.Topology) error {
+	tlsPolicy, ok := p.(*kuadrantv1alpha1.TLSPolicy)
+	if !ok {
+		return errors.New("invalid policy")
+	}
+
+	// Get all gateways that contains this policy
+	gws := lo.FilterMap(topology.Targetables().Items(), func(item machinery.Targetable, index int) (*machinery.Gateway, bool) {
+		gw, ok := item.(*machinery.Gateway)
+		return gw, ok && lo.Contains(gw.Policies(), p)
+	})
+
+	if len(gws) == 0 {
+		return errors.New("no valid gateways found")
+	}
+
+	for _, gw := range gws {
+		expectedCertificates := expectedCertificatesForGateway(ctx, gw.Gateway, tlsPolicy)
+
+		for _, cert := range expectedCertificates {
+			objs := topology.Objects().Children(gw)
+			obj, ok := lo.Find(objs, func(o machinery.Object) bool {
+				return o.GroupVersionKind().GroupKind() == CertManagerCertificateKind && o.GetNamespace() == cert.GetNamespace() && o.GetName() == cert.GetName()
+			})
+
+			if !ok {
+				return errors.New("certificate not found")
+			}
+
+			c := obj.(*controller.RuntimeObject).Object.(*certmanagerv1.Certificate)
+
+			conditions := utils.Map(c.Status.Conditions, func(c certmanagerv1.CertificateCondition) metav1.Condition {
+				return metav1.Condition{Reason: c.Reason, Status: metav1.ConditionStatus(c.Status), Type: string(c.Type), Message: c.Message}
+			})
+
+			if !meta.IsStatusConditionTrue(conditions, string(certmanagerv1.CertificateConditionReady)) {
+				return fmt.Errorf("certificate %s not ready", cert.Name)
+			}
+		}
+	}
+
+	return nil
+}
+
+func TLSPolicyAcceptedKey(uid types.UID) string {
+	return fmt.Sprintf("TLSPolicyValid:%s", uid)
+}
```

Comment:
Perhaps move this func to `tls_workflow.go`. It's used by the validator and the status updater, right?